CMakeLists.txt

```
@@ -0,0 +1,60 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+project(
+  CPM
+  VERSION 0.37.0
+  LANGUAGES NONE
+)
+
+option(CPM_INSTALL "Install CPM" OFF)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(
+  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPMConfig.cmake.in"
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfig.cmake"
+  INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CPM"
+)
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfigVersion.cmake" COMPATIBILITY SameMajorVersion
+  ARCH_INDEPENDENT
+)
+
+# For consumer using FetchContent
+if(NOT ${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
+  # Trick to use the find_package
+  configure_file(
+    "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPM.cmake" "${CMAKE_CURRENT_BINARY_DIR}/CPM.cmake" COPYONLY
+  )
+  set(CPM_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+  # Check if the user download the repo with git and not in an official release
+  if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git" AND IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+    find_package(Git REQUIRED QUIET)
+    # Generate a git-describe version string from Git repository tags.
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} tag --points-at HEAD
+      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+      OUTPUT_VARIABLE CPM_TAG
+      RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(NOT GIT_DESCRIBE_ERROR_CODE AND NOT CPM_TAG STREQUAL "")
+      set(CPM_RELEASE TRUE)
+    else()
+      set(CPM_RELEASE FALSE)
+    endif()
+  endif()
+  find_package(CPM CONFIG REQUIRED NO_POLICY_SCOPE)
+endif()
+
+if(CPM_INSTALL)
```

```suggestion
if (NOT CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```

This is sufficient, because we only need to test if it is called within ExternalProject, subdirectory, etc. All other cases it will not clash.

---

But why would they bundle `CPM.cmake` downstream when it will only be used for the current project? That can have unwanted licensing issues etc.

If we do want to allow that, then change the default values according to `CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME`

---

For example I have a repo that is compiling just a library A (it uses CPM). I have an other library B that depends on the other and use CPM too. The user can uses A without B but if he wants B it is simpler if he could use CPM to have B. It should save some doanloading and I could be sure user have CPM because I included in A. For the licensing issue I don't know

---

Note that `FetchContent` does not perform `install` when it is being used in the project, i.e. at `FetchContent_MakeAvailable`.

So if you are in project `C` that contains `A` which uses CPM to download `A.a` and `D`, and then you are in `B` which also uses CPM to download `B.b` and `D`. Then how CPM caches these so that we avoid re-downloading (`D`) will be on part of CPM implementation. On the `FetchContent` side (assuming `A` and `B` are using it to get CPM), CPM should be downloaded in the `C`'s [`${CURRENT_BINARY_DIR}/_deps`](https://cmake.org/cmake/help/latest/module/FetchContent.html#variable:FETCHCONTENT_BASE_DIR) even when called from `A` and `B`. I don't know how they deal with different versions if they hash it.

But still, would `CPM_INSTALL` help in this scenario? Because `B` and `C` will still download CPM regardless...

---

I mean the project A used CPM and A has been installed on host (with CPM)
User develops his own project and uses
```cmake
find_package(A)
find_package(CPM) # because he knows A has installed CPM

... CPM stufs to install B ... 
# no CPM download in A and B since B use CPM but CPM has already been found..

... User code

```

So CPM installed only once while it would be installed twice otherwise

---

`find_package` is primarily for finding system installed packages, it becomes very gray to call it like that. We can do it with `FetchContent` because there are `OVERWRITE_FIND_PACKAGE`, but otherwise it should have only such meaning.

But in this scenario, `C` already has CPM included (or installed) so there is no sense for `find_package(CPM)` (or rather, that was done before `find_package(A)`/`CPMAddPackage(A)`). Regardless if `B` calls `find_package(CPM)` or `FetchContent(CPM)` it should still cache the download to `C`'s defined location. It is on `B`'s implementation to properly use either `find_package` or `FetchContent` and it is not accessible to the user `C`.

Maybe I still don't get your scenario and would need a more detailed breakdown of it with how each `A`, `B`, `C` call `find_package`/`FetchContent` internally, and how `CPM_INSTALL` on `C` affects this.

---

Well, CMake now looks on packagename_ROOT variable when it exists as a path so CMake seems to officially allow this gray scenario.

For me find_package is a nice way to hide stufs from the consumer. Don't need to care where the package comes from as soon as the version is good.. you can check if package exists jas been install and if not do a fetcjcontent by yourself

---

I'm not talking about fetchcontent but the find_package :

From the cmake doc for find_package

New in version 3.12: Search paths specified in the <PackageName>_ROOT CMake variable and the <PackageName>_ROOT environment variable, where <PackageName> is the package to be found (the case-preserved first argument to find_package). The package root variables are maintained as a stack so if called from within a find module, root paths from the parent's find module will also be searched after paths for the current package. This can be skipped if NO_PACKAGE_ROOT_PATH is passed or by setting the CMAKE_FIND_USE_PACKAGE_ROOT_PATH to FALSE. See policy CMP0074.


Then recently find_package is able to fetchcontent fallback but i don't use this

As we required cmake 3.14 this policy is true per default so it's safe and if not it's because the consumer disavle it and it is in contradiction with what CPM wants so consumer should take care of the drawback

---

Oh, i see interesting but this is quite news... Would need to investigate but it should be fine because the find_package would find CPM as we provide one... If not it would fetchcontent it if you give i'm where to find.. 

This need to be check if there is an infinite loop but I imagine this as been thought

---

> This need to be check if there is an infinite loop but I imagine this as been thought

I don't think so. I haven't seen a project to find_package itself, so please test this case.

About the previous message. Note that no downstream package should `install(CPM)`, especially not in a way to be picked up by `find_package`. Upstream should be the ones that provide this package to Ubuntu, Fedora, etc. And with this PR we are actually able to ( Thanks :) ), once we finish polishing it.

---

Yes I agree what I want to do should be thought with caution but i think this PR looks fine for what people expect of a fetchcontent... CPM is special because you build nothing and just include module so including multy times is ok in CPM case it as been designed to it... 

I think we could finish the small polishing maybe supress the option and install only if CPM is the main project and donand other PR if necessary to include the option..

I really would like to bebable to dowload CPM with FetchContent it is much more convenient and safer than the get_cpm way

---

I agree with this plan :+1: